### PR TITLE
remove unused python39 from dockerfile

### PR DESCRIPTION
**Pull request by jbusche**
### Description of the change
Moving the dnf update to the end as I noticed it missed at least one package that was later installed.
Removing the old python3.9 items (and dnf) as we're using python3.11 and the legacy python3.9 is causing CVE violations. Attempts at upgrading python3.9 was causing many conflicts.

### Related issue number
Closes issue https://github.ibm.com/ai-foundation/watson-fm-stack-tracker/issues/828

### How to verify the PR
I did the following:

First I built the latest image from main to make sure it still had the CVEs... it does.

Then, I went through the following steps to test this PR on my OpenShift Cluster
1.  Clone the PR branch
```
git clone https://github.com/jbusche/fms-hf-tuning -b jb-cve-issue-828
cd fms-hf-tuning
```
2.  Build the image
```
docker build --progress=plain -t fms-hf-tuning:jim-0521-nocves . -f build/Dockerfile
```
3. Push it to my local OpenShift Repository
```
podman login -u kubeadmin -p $(oc whoami -t) $(oc registry info) --tls-verify=false
podman tag localhost/fms-hf-tuning:jim-0521-nocves  $(oc registry info)/opendatahub/fms-hf-tuning:jim-0521-nocves
podman push  --tls-verify=false $(oc registry info)/opendatahub/fms-hf-tuning:jim-0521-nocves
```
4. Launch a pytorchjob using that image with the following command:
```
cat <<EOF | kubectl apply -f -
apiVersion: v1
kind: ConfigMap
metadata:
  name: my-config
  namespace: default
data:
  config.json: |
    {
      "accelerate_launch_args": {
        "num_machines": 1,
        "num_processes": 2
      },
      "model_name_or_path": "bigscience/bloom-560m",
      "training_data_path": "/etc/config/twitter_complaints_small.json",
      "output_dir": "/tmp/out",
      "num_train_epochs": 1.0,
      "per_device_train_batch_size": 4,
      "per_device_eval_batch_size": 4,
      "gradient_accumulation_steps": 4,
      "evaluation_strategy": "no",
      "save_strategy": "epoch",
      "learning_rate": 1e-5,
      "weight_decay": 0.0,
      "lr_scheduler_type": "cosine",
      "logging_steps": 1.0,
      "packing": false,
      "include_tokens_per_second": true,
      "response_template": "\n### Label:",
      "dataset_text_field": "output",
      "use_flash_attn": false,
      "torch_dtype": "float32",
      "peft_method": "pt",
      "tokenizer_name_or_path": "bigscience/bloom"
    }
  twitter_complaints_small.json: |
    {"Tweet text":"@HMRCcustomers No this is my first job","ID":0,"Label":2,"text_label":"no complaint","output":"### Text: @HMRCcustomers No this is my first job\n\n### Label: no complaint"}
    {"Tweet text":"@KristaMariePark Thank you for your interest! If you decide to cancel, you can call Customer Care at 1-800-NYTIMES.","ID":1,"Label":2,"text_label":"no complaint","output":"### Text: @KristaMariePark Thank you for your interest! If you decide to cancel, you can call Customer Care at 1-800-NYTIMES.\n\n### Label: no complaint"}
    {"Tweet text":"If I can't get my 3rd pair of @beatsbydre powerbeats to work today I'm doneski man. This is a slap in my balls. Your next @Bose @BoseService","ID":2,"Label":1,"text_label":"complaint","output":"### Text: If I can't get my 3rd pair of @beatsbydre powerbeats to work today I'm doneski man. This is a slap in my balls. Your next @Bose @BoseService\n\n### Label: complaint"}
    {"Tweet text":"@EE On Rosneath Arial having good upload and download speeds but terrible latency 200ms. Why is this.","ID":3,"Label":1,"text_label":"complaint","output":"### Text: @EE On Rosneath Arial having good upload and download speeds but terrible latency 200ms. Why is this.\n\n### Label: complaint"}
    {"Tweet text":"Couples wallpaper, so cute. :) #BrothersAtHome","ID":4,"Label":2,"text_label":"no complaint","output":"### Text: Couples wallpaper, so cute. :) #BrothersAtHome\n\n### Label: no complaint"}
    {"Tweet text":"@mckelldogs This might just be me, but-- eyedrops? Artificial tears are so useful when you're sleep-deprived and sp\u2026 https:\/\/t.co\/WRtNsokblG","ID":5,"Label":2,"text_label":"no complaint","output":"### Text: @mckelldogs This might just be me, but-- eyedrops? Artificial tears are so useful when you're sleep-deprived and sp\u2026 https:\/\/t.co\/WRtNsokblG\n\n### Label: no complaint"}
    {"Tweet text":"@Yelp can we get the exact calculations for a business rating (for example if its 4 stars but actually 4.2) or do we use a 3rd party site?","ID":6,"Label":2,"text_label":"no complaint","output":"### Text: @Yelp can we get the exact calculations for a business rating (for example if its 4 stars but actually 4.2) or do we use a 3rd party site?\n\n### Label: no complaint"}
    {"Tweet text":"@nationalgridus I have no water and the bill is current and paid. Can you do something about this?","ID":7,"Label":1,"text_label":"complaint","output":"### Text: @nationalgridus I have no water and the bill is current and paid. Can you do something about this?\n\n### Label: complaint"}
    {"Tweet text":"Never shopping at @MACcosmetics again. Every time I go in there, their employees are super rude\/condescending. I'll take my $$ to @Sephora","ID":8,"Label":1,"text_label":"complaint","output":"### Text: Never shopping at @MACcosmetics again. Every time I go in there, their employees are super rude\/condescending. I'll take my $$ to @Sephora\n\n### Label: complaint"}
    {"Tweet text":"@JenniferTilly Merry Christmas to as well. You get more stunning every year \ufffd\ufffd","ID":9,"Label":2,"text_label":"no complaint","output":"### Text: @JenniferTilly Merry Christmas to as well. You get more stunning every year \ufffd\ufffd\n\n### Label: no complaint"}
---
apiVersion: "kubeflow.org/v1"
kind: PyTorchJob
metadata:
  name: ted-kfto-sft
  namespace: default
  labels:
    kueue.x-k8s.io/queue-name: lq-trainer
spec:
  pytorchReplicaSpecs:
    Master:
      replicas: 1
      restartPolicy: Never # Do not restart the pod on failure. If you do set it to OnFailure, be sure to also set backoffLimit
      template:
        spec:
          containers:
            - name: pytorch
              # This is the temp location util image is officially released
              image: image-registry.openshift-image-registry.svc:5000/opendatahub/fms-hf-tuning:jim-0521-nocves
              imagePullPolicy: IfNotPresent
              command:
                - "python"
                - "/app/accelerate_launch.py"
              env:
                - name: SFT_TRAINER_CONFIG_JSON_PATH
                  value: /etc/config/config.json
              volumeMounts:
              - name: config-volume
                mountPath: /etc/config
          volumes:
          - name: config-volume
            configMap:
              name: my-config
              items:
              - key: config.json
                path: config.json
              - key: twitter_complaints_small.json
                path: twitter_complaints_small.json
EOF
```
5. Make sure it completes successfully
```
oc get pytorchjobs,pods
NAME                                   STATE       AGE
pytorchjob.kubeflow.org/ted-kfto-sft   Succeeded   15m

NAME                        READY   STATUS      RESTARTS   AGE
pod/ted-kfto-sft-master-0   0/1     Completed   0          15m
```
6. Scan using twistlock the CVEs, and there are none.

### Was the PR tested
Yes, submitting using a pytorchjob and running the small twitter job above.

- [ ] I have added >=1 unit test(s) for every new method I have added.
- [ ] I have ensured all unit tests pass

### CODE REVIEW

**Comment by jbusche** [comment]
Main still has these CVEs:
```
cve | link | hasFix | status | packageType | packageName | packageVersion | packageLicense | packageBinaryPkgs | packagePath
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
CVE-2023-32681 | https://nvd.nist.gov/vuln/detail/CVE-2023-32681 | Y | fixed in 2.31.0 | python | requests | 2.25.1 | Apache 2.0 |   | /usr/lib/python3.9/site-packages/requests-2.25.1.dist-info
CVE-2024-35195 | https://nvd.nist.gov/vuln/detail/CVE-2024-35195 | Y | fixed in 2.32.0 | python | requests | 2.25.1 | Apache 2.0 |   | /usr/lib/python3.9/site-packages/requests-2.25.1.dist-info
CVE-2024-3651 | https://nvd.nist.gov/vuln/detail/CVE-2024-3651 | Y | fixed in 3.7 | python | idna | 2.1 | BSD-like |   | /usr/lib/python3.9/site-packages/idna-2.10-py3.9.egg-info
CVE-2022-40897 | https://nvd.nist.gov/vuln/detail/CVE-2022-40897 | Y | fixed in 65.5.1 | python | setuptools | 53.0.0 | UNKNOWN |   | /usr/lib/python3.9/site-packages/setuptools-53.0.0.dist-info
CVE-2023-45803 | https://nvd.nist.gov/vuln/detail/CVE-2023-45803 | Y | fixed in 2.0.7, 1.26.18 | python | urllib3 | 1.26.5 | MIT |   | /usr/lib/python3.9/site-packages/urllib3-1.26.5-py3.9.egg-info
CVE-2023-43804 | https://nvd.nist.gov/vuln/detail/CVE-2023-43804 | Y | fixed in 2.0.6, 1.26.17 | python | urllib3 | 1.26.5 | MIT |   | /usr/lib/python3.9/site-packages/urllib3-1.26.5-py3.9.egg-info
```

**Review by tedhtchang — APPROVED** [review]
/LGTM

File: build/Dockerfile [hunk 1]
```diff
@@ -112,11 +111,15 @@ ENV LIBRARY_PATH="$CUDA_HOME/lib64/stubs"
...
     && ln -s /usr/bin/python3.11 /bin/python \
     && python -m ensurepip --upgrade \
+    && dnf update -y \
     && dnf clean all
 
 # Removes the example private key to avoid high severity vulnerability warning
 RUN rm -f /usr/share/doc/perl-Net-SSLeay/examples/server_key.pem
 
+# Removes the python3.9 code to eliminate possible CVEs.  Also removes dnf
+RUN rpm -e $(dnf repoquery python3-* -q --installed) dnf python3 yum crypto-policies-scripts
```
Comment (anhuong) [inline build/Dockerfile:121]: this uninstalls just python3.9 and not all of python3 correct?

**Review by anhuong — APPROVED** [review]
i think if you've tested and it looks good, then LGTM

File: build/Dockerfile (hunk 1 above)
Comment (anhuong) [reply build/Dockerfile:121]: interesting that python --> python3.11 but python3 --> python3.9

Comment (jbusche) [reply build/Dockerfile:121]: Right, just checked, 
```
python --version
Python 3.11.7

python3.11 --version
Python 3.11.7
```

**Pull request merged**